tlvf: Add ClassList implementation

tomereli opened the pull request:
Currently, the CmduMessage class conflates the CMDU-specific things with
the general concept of a list of TLVs.

Create a new class ClassList that generalizes the handling of lists of
any TLVF generated classes (this includes lists of TLVs).

In a later commit/PR, this will be used in CmduMessage to represent the
list of TLVF classes that compose a CMDU - the CmduHeader and the list
of TLVs themselves.

In addition, this new container class will be used to represent inner
TLVs data such as vendor specific TLVs which contain a beerocks
operation header and class, and WSC TLVs which contains a list of
attributes.

All those will extend the ClassList to provide the parsing and creating
capabilities for the inner data.

Fixes #597

arnout committed:
framework: tlvf: add ClassList class

Currently, the CmduMessage class conflates the CMDU-specific things with
the general concept of a list of TLVs.

Create a new class ClassList that generalizes the handling of lists of
any TLVF generated classes (this includes lists of TLVs).

In a later commit/PR, this will be used in CmduMessage to represent the
list of TLVF classes that compose a CMDU - the CmduHeader and the list
of TLVs themselves.

In addition, this new container class will be used to represent inner
TLVs data such as vendor specific TLVs which contain a beerocks
operation header and class, and WSC TLVs which contains a list of
attributes.

All those will extend the ClassList to provide the parsing and creating
capabilities for the inner data.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Co-authored-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
common: tlvf: add beerocks_header class

Preperative commit.
Add beerocks_header header file which implements beerocks_header class
and derives from ClassList.
This will be used to replace the current beerocks vendor specific
handling and align it to use the inner Class lists approach according to

beerocks_header is not the best name for this class, since it is basically
a list of TLV like classes - first being the action header (cACTION_HEADER),
and the second is the operation itself.

A more appropriate name would have been beerocks_actions, but to keep
backward compatability, the name is kept (a comment is added in the code
in case we'll ever want to change the name.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout committed:
tlvf: add payload to vendor specific messages

Preparative commit.
Add payload to vendor specific messages, which will be used as the
buffer for the inner TLV list - beerocks_header.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: beerocks_message: use beerocks_header

Preparative commit.
Refactor add_intel_vs_data() to use the VS TLV payload and create a
beerock_header class on that buffer (reminder - beerocks_header derives
from TlvList).

The payload added is the maximum allowed TLV size which is MTU - CMDU
header length - EOM TLV length. On finalize of the whole CMDU, it will
resize the buffer to its real dimensions.

Also add get_beerocks_header which in followup commits will replace
get_vs_class.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: beerocks_message: use get_beerocks_header

Replace get_vs_class_header with get_beerocks_header.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: refactor parse_intel_vs_message

Refactor parse_intel_vs_message() to return beerocks_header, and align
all callers to the new return value.

Remove cmdu_rx from all APIs which now get beerocks_header since it
shouldn't be used.

While we're at it, make all Intel VS messages use
beerocks_header->addClass instead of cmdu_rx.addClass - using mechanical
substitution:

git grep -l 'cmdu_rx\.addClass<beerocks_message' | \
  xargs sed -i 's/cmdu_rx\.addClass<beerocks_message/beerocks_header->addClass<beerocks_message/g'

This replaced __most__ occurnces but not all (due to clang formatting
issues which pushed the '.addClass' to the next line), so also manually
update those.

This is a big change, and it is not possible to split it easilly but
most of it is mechanical and results from the API changes - this was
done intentionally since we wanted compilation to break so we'll replace
all the old usages of cmdu_rx.addClass with beerocks_header->addClass.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: Align verify_cmdu to VS TLV changes

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Refactor CmduMessage to use ClassList (breaks build)

Warning - this commit breaks the build. It is kept only for easier code
review of the feature as a whole.

---

Now that we have ClassList which implements a generic container of
tlvf-generated objects (classes), use it in CmduMessage:

- Now that we have a container for all types of tlvf classes, and since
cCMDU_HEADER is also a tlvf-generated class, simplify the CmduMessage
class implementation by having it always set as the first member in the
ClassList vector.

- Set and buff_len as constructor arguments since they should
never change.

- Implement forward wrappers to the ClassList for backward compatability
with the rest of the prplMesh components.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: refactor CmduMessageRx for ClassList (breaks build)

Now that CmduMessage uses ClassList, we need to refactor CmduMessageRx
accordingly to fix the build issues. This does not change any
functionality in the CmduMessageRx implementation.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: refactor CmduMessageTx to use ClassList (breaks build)

Now that CmduMessage uses ClassList, we need to refactor CmduMessageTx
accordingly to fix the build issues.
Since its going to be changed anyway with the new ClassList
implementation, remove the special handling for vendor specific TLVs in
the finalize function.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tree-wide: align to ClassList changes (fixes build)

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: socket thread: always swap on finalize

Preparative commit.

Always swap on finalize - we want to replace the finalize call so that
each BaseClass will implement its own finalize and will do the swap
there. The plan is to update addClass to always call finalize for the
previous class which was added. This will do the swapping always (we
don't have swap_needed in addClass), and will do the resizing if inner
TLVlist is present for that TLV - so the TLV finalize will call the
inner TLV finalize which will call finalize for every TLV in the inner
TLV vector and so forth.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tree-wide: remove swap_needed and m_swap

Previous commit removed the swap_needed from finalize and removed it
from the UDS header, so now we are effectively always doing swap on
finalize and parse, so now we can remove m_swap and swap_needed
everywhere which is what this commit is doing.
Followup commit will also remove m_swap from BaseClass - for now we
always set m_swap to true in addClass.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on common/beerocks/tlvf/src/src/beerocks_message.cpp:
```diff
@@ -164,7 +162,7 @@ bool message_com::send_cmdu(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
 
     uds_header->length      = cmdu_tx.getMessageLength();
-    uds_header->swap_needed = swap;
+    uds_header->swap_needed = true;
```
If it is always swapped then it seems to me that you can also remove the `swap_needed' from `uds_header`.

tomereli replied:
it happens on the followup commit

Commit:
tlvf: BaseClass: remove m_swap completely

Now that we __always__ swap, remove the unused m_swap member from
BaseClass and update all callers. Also update tlvf.py for all
autogenerated derived classes.

Followup commit will add the auogenerated files changes.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: update autogen files after m_swap removal

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on common/beerocks/tlvf/src/src/beerocks_message.cpp:
```diff
@@ -137,9 +137,7 @@ bool message_com::send_cmdu(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
-    bool swap = !dst_mac.empty();
-
-    if (!cmdu_tx.finalize(swap)) {
+    if (!cmdu_tx.finalize(true)) {
```
If the swap is always true, then maybe it is a good idea to remove the swap argument from the function.

tomereli replied:
happens in the followup commit

Commit:
tlvf: BaseClass: add finalize()

According to the tlvf ClassList redesign (#597), on every addClass we
will call finalize of the previous class, where finalize will include
sanity checks and any other operation that should be performed on
finilize such as endianity conversion.

This commit adds BaseClass::finalize() which does the post init check
and class swap, and updates ClassList::addClass() to call the finalize
on the previous class on every addClass.

Now that we do the finalize as we add classes to the vector, there is no
longer need to explicitely call swap() in ClassList::finalize() since
all classes but the last should already be swapped.

While we're at it - update CmduMessage::getMessageType() and
CmduMessage::getMessageId() to use the CmduHeader is_finalized() to
check if swapping is needed.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: BaseClass: add inner ClassList pointer

Every BaseClass can include an inner ClassList of other BaseClass
objects. This will be used for example for WSC TLV where the WSC
Attributes will be represented by an inner class list.

This commit adds the infrastructure for supporting inner class lists in
BaseClass by adding a new member, m_inner, which will be set when adding
an inner class list for a BaseClass object.

Since now a TLV may contain an inner class list, we need to call the
inner class list finalize when from the outer class finalize method.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: make BaseClass::finalize() pure virtual

Since we want BaseClass::finalize() to reduce the tailroom size of the
inner ClassList from both the m_buff_ptr__ of the outer class and the
length field in case it is a TLV class, it makes more sense to implement
finalize as pure virtual in BaseClass and have tlvf.py generate the
override function which calls the inner ClassList finalize (if such
exists), and reduces the tailroom (buffer length - message length) from
the current m_buff_ptr__ and updates m_length for TLV classes.

While we're at it - change the class_swap declaration in the generated
classes to include the override keyword, as we do now for finalize.

Also fix the tlvPrefilledData non-generated class since now it must
implement finalize().

Update to autogenerated files will be done in a separate commit.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: update autogenerated files for finalize changes

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on framework/tlvf/src/include/tlvf/ClassList.h:
```diff
@@ -0,0 +1,139 @@
+        return (const std::vector<std::shared_ptr<BaseClass>> &)m_class_vector;
+    }
+
+    template <class T> std::shared_ptr<T> addClass()
```
Since this one is here to stay, it might be good to add a doxygen comment now.

arnout commented on framework/tlvf/src/include/tlvf/ClassList.h:
```diff
@@ -0,0 +1,139 @@
+                ptr = std::make_shared<T>(m_buff, m_buff_len, m_parse, m_swap);
+            } else {
+                // do not allow to use addClass method if the previous tlv is not fully initialized
+                if (!m_class_vector.back()->isPostInitSucceeded()) {
```
Since we have this anyway, might be nice to change the code into:

```
auto prev = m_class_vector.back();
if (prev) {
   if (!prev->isPostInitSucceeded()) {
      ...
   }
   ptr = std::make_shared<T>(prev, m_parse, m_swap);
} else {
   ptr = std::make_shared<T>(m_buff, m_buff_len, m_parse, m_swap);
}
```

arnout commented on framework/tlvf/src/include/tlvf/CmduMessage.h:
```diff
@@ -20,133 +21,41 @@ namespace ieee1905_1 {
-protected:
-    void reset();
+    // Forward wrapper functions
+    // TODO check which of them can be removed
```
Good!

arnout commented on framework/tlvf/src/src/CmduMessage.cpp:
```diff
@@ -10,149 +10,46 @@
-    }
-
-    return int(tlvValue);
+    uint8_t *tlv = msg.getClassVector().back()->getBuffPtr();
```
I would like an explanatory comment here, saying that the first byte of the next TLV is the type.

More importantly, though: if we've consumed the entire buffer (so in practice, after the EOM TLV), getBuffPtr() may return a pointer beyond the buffer. So I think we should still check getBuffRemainingBytes explicitly. I would actually keep the original condition:

```
if (prev->getBuffRemainingBytes() < kTlvHeaderLength) {
```

Also it would be nicer to (in a preparatory commit) add a private header function that returns a sTlvHeader and that can be used by all of the below functions.

arnout commented on framework/tlvf/src/src/CmduMessage.cpp:
```diff
@@ -10,149 +10,46 @@
 {
-    uint8_t tlvValue = 0;
-    if (!m_cmdu_header && (m_cmdu_header->getBuffRemainingBytes() < kTlvHeaderLength)) {
+    if (!getCmduHeader())
```
I think it will be a lot easer to understand what's going on if we make use of the ClassList:

```
auto prev = msg.getClassVector().back();
if (!prev)
    return -1;
```

(prev in a variable because I'll use it again below.)

Also, I think the msg.getClassVector().back() is going to be used so often that it pays of to add a method for it in ClassList.

arnout commented on framework/tlvf/src/src/CmduMessage.cpp:
```diff
@@ -10,149 +10,46 @@
+    return *tlv;
 }
 
 bool CmduMessage::getNextTlvType(eTlvType &tlvType) const
```
AFAICS this function only gets used in tlvf_test, so maybe we should just remove it in a preparatory commit...

tomereli replied:
Will be addressed in #601, hence leaving open.

arnout commented on framework/tlvf/src/include/tlvf/CmduMessageRx.h:
```diff
@@ -22,14 +22,14 @@ class CmduParser;
+    CmduMessageRx(uint8_t *buff, size_t buff_len);
     ~CmduMessageRx();
-    bool parse(uint8_t *buff, size_t buff_len, bool swap_needed = true, bool parse_tlvs = false);
+    bool parse(bool swap_needed = true, bool parse_tlvs = false);
```
This change (moving the buffer from parse to constructor) could be a separate, preparatory commit. It is relatively easy to make that change and change the calls to parse() and the constructors in the same commit.

tomereli replied:
I tried doing it in a separate commit but it became a spaghetti of changes so decided not to do that eventually.

arnout commented on framework/tlvf/src/include/tlvf/CmduMessageRx.h:
```diff
@@ -22,14 +22,14 @@ class CmduParser;
 
 private:
-    std::list<std::shared_ptr<CmduParser>> parsers_;
+    std::shared_ptr<CmduParser> parser;
```
This change is definitely easy to do in a separate, preparatory commit.

tomereli replied:
Yes, but at this point what is there to gain from it? I prefer not to block merge for this PR for small things like this one since every new change that enters prplMesh may break this one (we'll need to find all those places we need to replace `cmdu_tx.addClass()` with `beerocks_header->addClass()`)

arnout commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -12,47 +12,26 @@
         return false;
-    }
+    if (swap_needed)
+        cmduhdr->class_swap(); //swap back cmduheader
```
I think this is the only change that is directly related to the ClassList, so it can easily be done together with the previous commit.

tomereli replied:
https://github.com/prplfoundation/prplMesh/pull/584#discussion_r360714833

arnout commented on framework/tlvf/src/src/CmduMessageTx.cpp:
```diff
@@ -13,62 +13,39 @@
-    if (!m_buff || !m_buff_len)
+    // parse - false since this is for TX
+    // swap - false since on TX we do swap in Finalize(swap_needed)
+    msg.reset(false, false);
```
This could also be done in a preparatory commit: add a protected reset function to CmduMessage which has the parse and swap arguments, and use that instead of setting parse and swap explicitly. And since we always do an addClass<cCmduHeader> afterwards (or a std::make_shared in the original code), we can move that into the reset function as well. And the memset should be moved into the function as well, of course.

tomereli replied:
https://github.com/prplfoundation/prplMesh/pull/584#discussion_r360714833

arnout commented on framework/tlvf/src/src/CmduMessageTx.cpp:
```diff
@@ -13,62 +13,39 @@
 
     auto tlv = addClass<tlvVendorSpecific>();
-    if (!tlv) {
-        m_cmdu_header.reset();
```
This reset is removed now. Probably it shouldn't be done (we don't automatically reset in other cases), but better in a separate, explicit commit.

tomereli replied:
https://github.com/prplfoundation/prplMesh/pull/584#discussion_r360714833

arnout commented on framework/tlvf/src/src/CmduMessageTx.cpp:
```diff
@@ -77,47 +54,7 @@ std::shared_ptr<tlvVendorSpecific> CmduMessageTx::add_vs_tlv(tlvVendorSpecific::
-        }
-    }
-
-    // Update vendor specific length which is seperated to 3 classes
-    for (auto class_it = m_class_vector.begin(); class_it != m_class_vector.end(); class_it++) {
-        // Type is the first byte in the tlv buffer
-        uint8_t tlv_type = *(*class_it)->getStartBuffPtr();
-        if (tlv_type == uint8_t(eTlvType::TLV_VENDOR_SPECIFIC)) {
-            uint16_t *tlv_length = reinterpret_cast<uint16_t *>(
-                (class_it->get())->getStartBuffPtr() + sizeof(uint8_t));
-            // tlv vendor specific length already include OUI and beerocks header length
-            // need to add only beerocks message length.
-            // Classes order: +0 = tlvVendorSpecific, +1 = beerocks_header, +2 = beerocks_message
-            if (std::next(class_it, 1) == m_class_vector.end() ||
-                std::next(class_it, 2) == m_class_vector.end()) {
-                TLVF_LOG(ERROR) << "vendor specific TLV has not been added correctly";
-                return false;
-            }
-            *tlv_length += (std::next(class_it, 2)->get())->getLen();
-            class_it += 2;
-        }
-    }
```
I'm not convinced that it's a good idea to remove this. This hack should still just work with the ClassList.

tomereli replied:
I don't understand why we should keep this, but keeping open so we can review it again in #601

arnout commented on framework/tlvf/src/src/CmduMessageTx.cpp:
```diff
@@ -77,47 +54,7 @@ std::shared_ptr<tlvVendorSpecific> CmduMessageTx::add_vs_tlv(tlvVendorSpecific::
-    return true;
-}
+    return msg.finalize(swap_needed);
+};
```
Missing newline.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:
```diff
@@ -0,0 +1,50 @@
+ */
+class beerocks_header : public ClassList {
+public:
+    beerocks_header() = delete;
```
Since the base class ClassList already deleted it, this is implied (nitpick).

tomereli replied:
I think it is still good practice to delete this explicitly.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:
```diff
@@ -0,0 +1,50 @@
+    }
+    std::shared_ptr<beerocks_message::cACTION_HEADER> actionhdr()
+    {
+        return getClass<beerocks_message::cACTION_HEADER>();
```
Would it make sense to instead do a dynamic_cast on the front() of the class vector?

tomereli replied:
I think not, better to keep it generic by using `getClass` to make sure we cover all corner cases.

arnout commented on framework/tlvf/tlvf.py:
```diff
@@ -1556,16 +1555,16 @@ def addClassSizeMethod(self, insert_name, insert_marker, name):
-            self.insertLineH(obj_meta.name, self.CODE_CLASS_CONSTRACTOR, "%s(uint8_t* buff, size_t buff_len, bool parse = false, bool swap_needed = false);" % (obj_meta.name))
-            self.insertLineCpp(obj_meta.name, self.CODE_CLASS_CONSTRACTOR, ("%s::%s(uint8_t* buff, size_t buff_len, bool parse, bool swap_needed) :" % (obj_meta.name, obj_meta.name)))
-            self.insertLineCpp(obj_meta.name, self.CODE_CLASS_CONSTRACTOR, ("%sBaseClass(buff, buff_len, parse, swap_needed) {" % self.getIndentation(1) ))
+            self.insertLineH(obj_meta.name, self.CODE_CLASS_CONSTRACTOR, "%s(uint8_t* buff, size_t buff_len, bool parse = false);" % (obj_meta.name))
```
framework/tlvf/README.md still refers to the swap_needed argument.

There's also one in son_slave_thread.cpp comment and in beerocks_message.cpp.

tomereli replied:
Will be fixed as part of #601

Commit:
tlvf: BaseClass: make destructor default virtual

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on framework/tlvf/src/src/BaseClass.cpp:
```diff
@@ -41,3 +43,22 @@ bool BaseClass::buffPtrIncrementSafe(size_t length)
+        return false;
+    }
+    class_swap();
+    m_finalized = true;
```
Hm, this is risky...

Before, the m_swapped field of ClassList was checked in finalize and it was not done again if it was already swapped. Now, only the finalized flag is checked. Thus, the following sequence would have worked previously but no longer works: `cmdu.swap(); cmdu.finalize();` And there are probably more corner cases.

I think it would be good to do a review of all explicit cmdu.swap and foo.class_swap calls and explain why they're OK (for example, in a reply to this comment). And of course testing.

tomereli replied:
Well, it passed all tests so I guess it's fine. Nevertheless - to be reviewed again as part of #601, so keeping open.

arnout commented on framework/tlvf/src/src/BaseClass.cpp:
```diff
@@ -41,3 +43,22 @@ bool BaseClass::buffPtrIncrementSafe(size_t length)
+    class_swap();
+    m_finalized = true;
+    return true;
+}
```
Missing newline.

tomereli replied:
Fixed

arnout reviewed (commented):
Not finished yet but already posting my comments.

Mostly it's about better ways to split and order the commits or nitpick comments. The only important ones so far are:

- check that getNextTlvType and friends can't read beyond the buffer;
- check the explicit swap calls.

arnout commented on framework/tlvf/src/include/tlvf/ClassList.h:
```diff
@@ -11,10 +11,10 @@
 #define _ClassList_H_
 
 #include <memory>
-#include <tlvf/BaseClass.h>
```
There's no need to remove this include.

arnout commented on framework/tlvf/src/include/tlvf/ClassList.h:
```diff
@@ -11,10 +11,10 @@
-#include <tlvf/ieee_1905_1/eTlvType.h>
 #include <vector>
 
+class BaseClass;
```
There's no need for a forward declaration here.

arnout commented on framework/tlvf/src/include/tlvf/ClassList.h:
```diff
@@ -36,7 +36,7 @@ class ClassList {
             } else {
                 // before adding a new class, finalize the previous one
-                if (!m_class_vector.back()->finalize())
+                if (!finalize(m_class_vector.back()))
```
There's no need for this weird finalize function.

arnout commented on framework/tlvf/src/include/tlvf/BaseClass.h:
```diff
@@ -42,7 +42,7 @@ class BaseClass {
     const size_t m_buff_len__;
     const bool m_parse__;
-    bool m_finalized;
+    bool m_finalized__;
```
Why is this rename needed?

tomereli replied:
To align with other variable names which are explicitly used in tlvf.py so we use the __ suffix to mark that.

arnout commented on framework/tlvf/src/src/BaseClass.cpp:
```diff
@@ -44,24 +44,3 @@ bool BaseClass::buffPtrIncrementSafe(size_t length)
-    class_swap();
-    m_finalized = true;
-    return true;
-}
```
I would prefer to keep this function in BaseClass, and add a private virtual finalize_impl function or something like that that is optionally generated by tlvf and that defaults to nothing. That keeps most of the code outside of tlvf which is nicer.

tomereli replied:
Will be reviewed as part of #601

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -61,30 +61,48 @@ class message_com {
+        const size_t max_vs_tlv_length = ieee1905_1::CmduMessage::kMaxCmduLength -
+                                         ieee1905_1::CmduMessage::kCmduHeaderLength -
+                                         ieee1905_1::tlvVendorSpecific::get_initial_size() -
+                                         ieee1905_1::tlvEndOfMessage::get_initial_size();
```
I still don't understand why we don't just allocate the rest of the buffer here. That's basically what we've done up to now and it seems to work...

tomereli replied:
Didn't cherry pick the commit yet, we will

Commit:
tlvf: cleanups: remove getClass

Cfr
https://github.com/prplfoundation/prplMesh/pull/551#discussion_r356576428

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on framework/tlvf/src/include/tlvf/ClassList.h:
```diff
@@ -0,0 +1,139 @@
+    {
+        std::shared_ptr<T> ptr;
+        if (m_buff) {
+            if (m_class_vector.size() == 0) {
```
m_class_vector.empty() (nitpick)

morantr replied:
Please remove nesting as most as possible by handling an error right away and avoid `else`, it makes the code much more readable and clearer:
```cpp
template <class T> std::shared_ptr<T> addClass()
{
    std::shared_ptr<T> ptr;
    if (m_buff == nullptr) {
        return ptr;
    }
        
    if (m_class_vector.size() == 0) {
        ptr = std::make_shared<T>(m_buff, m_buff_len, m_parse, m_swap);
    } else {
        // do not allow to use addClass method if the previous tlv is not fully initialized
        if (!m_class_vector.back()->isPostInitSucceeded()) {
            TLVF_LOG(ERROR) << "TLV post init failed";
            return ptr;
        }
        ptr = std::make_shared<T>(m_class_vector.back(), m_parse, m_swap);
    }

    if (!ptr || ptr->isInitialized() == false) {
        return std::shared_ptr<T>();
    }
    
    m_class_vector.push_back(std::shared_ptr<BaseClass>(ptr));

    return std::shared_ptr<T>();
}
```

tomereli replied:
Done

morantr commented on framework/tlvf/src/include/tlvf/CmduMessage.h:
```diff
@@ -20,133 +21,40 @@ namespace ieee1905_1 {
-        }
-        return ptr;
-    }
+    ClassList msg;
```
I don't understand why did you make `ClassList` be a member of `CmduMessage` instead of just be it.
I mean you took out all kinds of functions out from `CmduMessage` to `ClassList`, why couldn't you just change `CmduMessage`?

Also, why did you choose to use a member over heritage? It would have save you all the wrapper functions.

tomereli replied:
>I don't understand why did you make ClassList be a member of CmduMessage instead of just be it.

Don't understand what you mean by "just be it." - the __whole__ point in this change is to decouple the TLVs from CmduMessage class, which is why we crearted `ClassList` and put it as a member of `CmduMessage`.
It also makes perfect sense - a Cmdu Message is a descriptor of TLVF generated classes, so it includes a member ClassList which holds all the TLVs, where the first class is the CmduHeader and the rest are the TLVs themselves.

>Also, why did you choose to use a member over heritage? It would have save you all the wrapper functions.

The wrapper functions are only for backward compatibility. We chose member over inheritance cause it makes more sense and we don't want CmduMessage to access all the the private and protected members of ClassList.

morantr commented on framework/tlvf/src/src/CmduMessage.cpp:
```diff
@@ -10,149 +10,46 @@
+    if (!getCmduHeader())
+        return -1;
+    uint8_t *tlv = msg.getClassVector().back()->getBuffPtr();
+    return *(uint16_t *)(tlv + sizeof(uint8_t));
```
For pointer casting, please you `reinterpret_cast`

tomereli replied:
Done

morantr commented on framework/tlvf/src/src/CmduMessageTx.cpp:
```diff
@@ -13,62 +13,39 @@
-
-    return m_cmdu_header;
+    msg.reset(true, false);
+    return addClass<cCmduHeader>();
```
We talked about it. This function implementation is not correct. Add a comment so you could resolve it when you will finish with cmdu copy constructor.

tomereli replied:
No need for the copy constructor. The latest works - you are looking at a broken commit.
Since this change is so big, I had to include some commits which break things so it would be reviewable. I agreed with Arnout to keep the broken commits so in the future it would be easy to review this change for someone who is not familiar with TLVF.

morantr commented on framework/tlvf/src/include/tlvf/CmduMessageRx.h:
```diff
@@ -22,14 +22,14 @@ class CmduParser;
+    CmduMessageRx(uint8_t *buff, size_t buff_len);
     ~CmduMessageRx();
-    bool parse(uint8_t *buff, size_t buff_len, bool swap_needed = true, bool parse_tlvs = false);
+    bool parse(bool swap_needed = true, bool parse_tlvs = false);
```
Why did you remove the `uint8_t *buff, size_t buff_len` from the parse function and move it to the constructor?
What does is the benefit of that? I don't like it since it would force calling the constructor in class initialize list if cmdu_rx is a member of a class.

tomereli replied:
It makes perfect sense to move the stuff which should never change to the constructor, such as the buffer and buffer length. The benefit is that this is the only way I could easilly integrate ClassList to CmduMessage - otherwise I would have had to do some pretty ugly adaptations whenever create / parse is called.

morantr commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/tlvVendorSpecific.h:
```diff
@@ -39,8 +40,11 @@ class tlvVendorSpecific : public BaseClass
         
         const eTlvType& type();
-        uint16_t& length();
+        const uint16_t& length();
```
What is the point of passing the length as const reference? We pass const reference only if the size is greater than 4 bytes. which is not the case here. (nitpick)

https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#functions

tomereli replied:
This is autogenerated code, this means that the length field should never change. We only pass references in tlvf, so const is needed here.

morantr commented on common/beerocks/bcl/source/beerocks_socket_thread.cpp:
```diff
@@ -209,9 +209,7 @@ bool socket_thread::verify_cmdu(message::sUdsHeader *uds_header)
-                auto beerocks_magic = *(
-                    uint32_t *)((uint8_t *)tlv + ieee1905_1::tlvVendorSpecific::get_initial_size());
-
+                auto beerocks_magic = *(uint32_t *)(tlv_vendor_specific.payload());
```
We should get rid of the magic in the beerocks action header. It is not relevant anymore, and it is basically a leftover of the area before we used cmdu.

tomereli replied:
It still works and it still validates the TLV, so I don't think removing it now is beneficial.
Removing this will be done as part of moving verify_cmdu to the transport.

Commit:
tlvf: add struct TlvHeader

Preparative commit.
Move it from bcl to tlvf where it belongs so it can also be used by
TLVF.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on framework/tlvf/src/include/tlvf/CmduMessage.h:
```diff
@@ -52,10 +53,21 @@ class CmduMessage {
+    // However, when the CMDU is sent internally via UDS, this restriction does not apply.
+    // So we allo changing it via set_max_tlv_length() API.
+    size_t m_max_tlv_length =
+        kMaxCmduLength - kCmduHeaderLength - ieee1905_1::tlvEndOfMessage::get_initial_size();
```
I'm sorry but this whole commit is featuring a bad solution IMO. Limiting CMDU size to MTU and creating special cases for "jumbo" tlvs is not an elegant solution.
There should not be any size limit expect the defined by the given buffer size, which is:
`buff_len - cmdu_header_len - tlv_end_of_message_len'

Limiting the TLV size should not be handled in the tlvf, it should be handled in the transport layer.
If we have exceeded MTU size, we should get that failure on the BTL or the transport when trying to send the CMDU.

tomereli replied:
The ieee1905 defines fragmantation to be based on TLV boundary, so each TLV must never exceed the MTU - CMDU header length.
Adding the check here is also necessary since we don't know if this CMDU will be transmitted over the transport.

> Limiting the TLV size should not be handled in the tlvf, it should be handled in the transport layer.

Correct - and it is handled there, which is how this was reproduced in the first place - there is an error in the transport when trying to send oversized TLVs - see `ieee1905_transport_packet_processing.cpp::fragment_and_send_packet_to_network_interface()` function.

We can't know in advance if the TLV should be sent over to the network, in which case it will fail if it has an oversized TLV. The best solution is never to send a TLV which is larger then `MTU - CMDU header length - TLV end of message length` - but this is problematic since `network_map` minimum node size is already bigger than 2KB.

The solution of using the jumbo TLVs is the most elegant I could think of, I'm sorry you don't like it, but I think that for now we should take it as is since this PR is already huge and took too long to be merged, we can refactor that in the future if we decide to do so.

morantr replied:
> We can't know in advance if the TLV should be sent over to the network, in which case it will fail if it has an oversized TLV. The best solution is never to send a TLV which is larger then MTU - CMDU header length - TLV end of message length - but this is problematic since network_map minimum node size is already bigger than 2KB.

You don't need to know it in advance. You should get an error when trying to send such TLV to the bus. So if such an error exists on the transport, then the only change you can do in this PR is to set the tlv limitation to `buf_len - CMDU header length - TLV end of message length` and remove the mumbo jumbo commits.

tomereli replied:
Moran - Well done, I just misunderstood you but now with the resize solution in place there is no longer any need to limit the TLV size, so I removed the relevant commits. Thanks for finding that!

https://media.giphy.com/media/vCKC987OpQAco/giphy.gif

Commit:
tlvf: CmduMessage use sTlvHeader

Cfr https://github.com/prplfoundation/prplMesh/pull/584#discussion_r359954976

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
